src/main/java/com/flightstats/hub/dao/aws/S3WriteQueue.java

```diff
@@ -14,57 +13,62 @@
 import com.github.rholder.retry.StopStrategies;
 import com.github.rholder.retry.WaitStrategies;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
-import com.google.inject.Inject;
 import com.google.inject.Singleton;
 import com.google.inject.name.Named;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import lombok.extern.slf4j.Slf4j;
 
+import javax.inject.Inject;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 
-@SuppressWarnings("WeakerAccess")
 @Singleton
+@Slf4j
 public class S3WriteQueue {
 
-    private final static Logger logger = LoggerFactory.getLogger(S3WriteQueue.class);
+    private final Retryer<Void> retryer = buildRetryer();
 
-    private static final int THREADS = HubProperties.getS3WriteQueueThreads();
-    private static final int QUEUE_SIZE = HubProperties.getS3WriteQueueSize();
-    private Retryer<Void> retryer = buildRetryer();
-    private BlockingQueue<ChannelContentKey> keys = new LinkedBlockingQueue<>(QUEUE_SIZE);
-    private ExecutorService executorService = Executors.newFixedThreadPool(THREADS,
-            new ThreadFactoryBuilder().setNameFormat("S3WriteQueue-%d").build());
-    @Inject
-    @Named(ContentDao.WRITE_CACHE)
-    private ContentDao spokeWriteContentDao;
-    @Inject
-    @Named(ContentDao.SINGLE_LONG_TERM)
-    private ContentDao s3SingleContentDao;
-    @Inject
-    private StatsdReporter statsdReporter;
+    BlockingQueue<ChannelContentKey> keys;
+    private ExecutorService executorService;
+
+    private final WriteQueueConfig writeQueueConfig;
+    private final ContentDao spokeWriteContentDao;
+    private final ContentDao s3SingleContentDao;
+    private final StatsdReporter statsdReporter;
 
     @Inject
-    private S3WriteQueue() throws InterruptedException {
-        logger.info("queue size {}", QUEUE_SIZE);
-        for (int i = 0; i < THREADS; i++) {
+    S3WriteQueue(@Named(ContentDao.WRITE_CACHE) ContentDao spokeWriteContentDao,
+                 @Named(ContentDao.SINGLE_LONG_TERM) ContentDao s3SingleContentDao,
+                 StatsdReporter statsdReporter,
+                 WriteQueueConfig writeQueueConfig) {
+        this.spokeWriteContentDao = spokeWriteContentDao;
+        this.s3SingleContentDao = s3SingleContentDao;
+        this.statsdReporter = statsdReporter;
+        this.writeQueueConfig = writeQueueConfig;
+        keys = new LinkedBlockingQueue<>(writeQueueConfig.getQueueSize());
+        executorService = Executors.newFixedThreadPool(writeQueueConfig.getThreads(),
+                new ThreadFactoryBuilder().setNameFormat("S3WriteQueue-%d").build());
+    }
+
+    public void start() throws InterruptedException {
```

Comment:
I think this should be refactored into an AbstractService https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/util/concurrent/AbstractService.html 
We are using that pattern throughout the rest of the hub for similar long running "services"
Would like feedback before I went further on that.

Reply:
I like that idea, especially if we break with the pattern of doing it as a nested class, if possible.